### ISSUE TITLE
maybe we should drop Zope Interface

### ISSUE THREAD

**Issue by glyph** [twisted/twisted#12364]
We currently use zope.interface which adds some layers of complexity. Just off the top of my head:

- It requires that we use a mypy plugin to check its types, and therefore editing Twisted code does not behave correctly in other type checkers.  This means that editor support for Twisted code is reduced out of the box.
  - This also means that we are very limited in what we can express, because zope.interface itself and the typing plugin are missing [important features like generics](https://github.com/Shoobx/mypy-zope/issues/95)
- It contains a native acceleration module which becomes another blocker for trying new Python features like free-threading
- It uses [setuptools' deprecated namespace packges](https://github.com/zopefoundation/zope.interface/issues/270), which requires a runtime dependency on setuptools
- the lack of 'self' on interface declarations is confusing both to tooling and to new developers
- because of all these complexities, bootstrapping even targeted subsets of Twisted in environments like Micropython or PyScript becomes more complex and annoying

In return, we get:

1. [nominal abstract types](https://blog.glyph.im/2021/03/interfaces-and-protocols.html)
2. [a very simple adapter registry](https://zopeinterface.readthedocs.io/en/latest/human.html), that we actually have to [implement ourselves](https://github.com/twisted/twisted/blob/a6c0e2dd1df7f7610aa4efe64a442db829954e44/src/twisted/python/components.py) out of a small number of parts, which we [hardly ever use](https://github.com/search?q=repo%3Atwisted%2Ftwisted+registerAdapter%28+-path%3Adocs%2F+-path%3A%2Ftest%2F&type=code)

When we first adopted zope.interface, decades ago, this was a pretty good tradeoff, but as Python's native facilities have improved, the tradeoff has gotten worse and worse.

Oddly, we still can't get nominal abstract types from native Python [due to this bug](https://github.com/python/mypy/issues/2922).

Adopting type annotations has shown us that a lot of our uses of adaptation are actually superfluous or unnecessary, and with a-priori knowledge of types, we can mostly forego it. We do still have a few places like twisted.web.template and twisted.cred which actually have sensible uses for such a thing, but our uses could mostly be satisfied by a few dictionary lookups with `Protocol` objects.

I think we might want to consider a migration away from zope.interface towards just using `typing.Protocol` for everything.  The compatibility migration here will be an interesting puzzle, but one that I think is eminently solvable; most of the interesting uses of zope.interface are at runtime.  We can probably start by switching to declaring our interfaces *as* protocols and then having a runtime shim that constructs zope.interface objects so that things like adaptation still work, and then making the dependency optional.

**Comment by glyph** [comment]
> * because of all these complexities, bootstrapping even targeted subsets of Twisted in environments like Micropython or PyScript becomes more complex and annoying

As an example of how this sort of thing can bite us in unforseen and unforseeable interactions of this sort of complexity, @alex prompted me to write this bug because he is currently in an environment with a security configuration which gets angry about `.exe` files in python packages.  Setuptools contains `.exe` stubs, which prevents installation of Twisted, because Twisted has a runtime dependency on Setuptools, because of zope.interface (and because of incremental, but one thing at a time).

Is there anything really technically *wrong* with a runtime dependency on setuptools? No, this is arguably a misconfiguration and a weird poorly-motivated threat model where .exe files are "bad" but python packages are "ok". Nevertheless, this sort of logic exists everywhere, and we end up dragging these native binary executable files around just because zope.interface has the baggage of the "namespace package" that predates PEP 420 (which has been around for 12 years now) which, as linked above, they are having difficulty switching away from even now.  If we were getting a lot of value out of this dependency for most of our users I'd say that we could safely ignore it, but the dependency is so tenuous that it is almost accidental, and is delivering nearly nothing.

**Comment by twm** [comment]
How would we make this change in a backward-compatible manner?

**Comment by glyph** [comment]
Same way we do anything else; make a new thing, deprecate the old thing, remove the old thing.

More specifically:

- Define a new function that can do adaptation on `Protocol` instances.
- Create a utility function that derives a runtime `Interface` from a `Protocol`.
- For every `interfaces` module, Make a new module, let's say, `twisted.internet.boundaries` for `twisted.internet.interfaces`; redefine every `Interface` as a `Protocol`.  (It will probably need to be `@runtime_checkable` for various do-you-support-this checks.)
- Define runtime aliases in `twisted.internet._interfacecompat`, which everything in Twisted can import for `@implementer` decorations; re-export them in `twisted.internet.interfaces`.
- Update the documentation to remove mention of interfaces; publish a brief migration guide.
- Deprecate `twisted.internet.interfaces` (and similar modules).
- Wait.
- Remove `twisted.internet.interfaces` (and similar modules).

There are some edge cases and potentially some performance concerns, but nothing beyond a simple matter of programming.

**Comment by glyph** [comment]
There are also bits of runtime trickery that we can use to allow the compatibility interfaces to only be defined if you import the compatibility namespace.  Notably we'd need our own `@implementer` for that, which could conditionally import the underlying `@implementer`. We would probably want this so you can use Twisted without the zope.interface dep *before* the deprecation cycle completed.

**Comment by adiroiban** [comment]
I expect that there will be a lot of work...

but one thing that we can do is avoid using zope.interface for any new code.

I am +1 for removing zope.interface

**Comment by tristanlatr** [comment]
A note on the documentation: pydoctor does not support structural subtyping at the moment. 

It will be quite interesting and probably difficult to migrate the support we currently have for interfaces toward a Protocol based system. Pydoctor does not carry the complexity to do proper type inference for now, and thus won’t always be able to recognize a certain class as being an implementation of a certain protocol (if it’s not explicitly listed in the base classes). 

We can probably do something like checking if the required symbols of a protocol are defined, let that be a callable we can check the compatibility of the number of arguments, but we won’t be able to verify the symbol types. This process will probably be time consuming as well.

**Comment by glyph** [comment]
> but one thing that we can do is avoid using zope.interface for any new code.

I don't think that this actually helps:

1. the marginal effort in the cutover (the amount of work per interface) is pretty small, so avoiding new uses makes the code inconsistent
2. when we are done designing our new APIs (with our new adaptation tool, etc) it'll be easier to switch from consistent use of zope.interface to consistent use of the new thing than to fix up one-off uses of `isinstance` that we have cooked up in the meantime

**Comment by glyph** [comment]
> It will be quite interesting and probably difficult to migrate the support we currently have for interfaces toward a Protocol based system. Pydoctor does not carry the complexity to do proper type inference for now, and thus won’t always be able to recognize a certain class as being an implementation of a certain protocol (if it’s not explicitly listed in the base classes).

Good point.  One of my longest-standing frustrations with mypy was that it was impossible to do this:


```python3
from dataclasses import dataclass
from typing import TypeVar, Protocol, Generic, Callable

T = TypeVar("T")


@dataclass
class conformer(Generic[T]):
    proto: Callable[..., T]

    def __call__(self, impl: type[T]) -> type[T]:
        return impl


class P(Protocol):
    def m(self) -> None:
        "method"

    # def q(self) -> None:
    #     "other"

@conformer(P)
class C:
    def m(self) -> None:
        "implementation"
```

But… now… maybe it is!? If I `reveal_type(C())` I get what I expct, if I uncomment the definition of `q` I get a type failure as I should…

So, @tristanlatr, could we add a micro-library to support Pydoctor that has… basically just that, in it, which we can then use to declare explicit conformance with relevant `Protocol`s to get them to show up in the docs?

**Comment by tristanlatr** [comment]
@glyph this is a nice proposal and it worth investigating it further. But I don't think putting this helper in a micro-library is the way to go. If it's really useful it should be discussed for inclusion in the `typing.py` module.

I've just read about [@runtime_checkable](https://docs.python.org/3/library/typing.html#typing.runtime_checkable) decorator and I've learned that it does not even try to match callable signatures it only checks for the presence of names, more, a recent changes in python 3.12 makes dynamically assigned attributes ignored by the checks. So I think we can have the same kind of functionality in pydoctor, it's just that the implementation will be tricky (because in order to avoid exponential complexity we'll have to reason about both protocols and classes hierarchies and come up with an algorithm that computed all protocol matches in the system in the minimum number of steps - but this is another discussion).

**Comment by glyph** [comment]
> If it's really useful it should be discussed for inclusion in the `typing.py` module.

This is going to be a huge, slow, and extremely contentious process.  I am thinking of a library specifically to include the documentation hint, "this should be documented as complying with this other protocol".

I have no objection to proposing this upstream, I just don't think we should wait the 3-5 years required to get it included before we start on this project.
